File: src/functions/scraper.js
import puppeteer from 'puppeteer';

import { CONSTANTS } from '../utils/constants.js';

export const scraper = async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();

  await page.goto(CONSTANTS.appURL, {
    waitUntil: 'networkidle2',
    timeout: 60000,
  });

  const data = await page.evaluate(async CONSTANTS => {
    const categories = Array.from(
      document.querySelectorAll(CONSTANTS.categoriesSectionClass),
    );

    return categories.map(category => {
      const element = category.querySelector(CONSTANTS.categoryTitleClass);
      const elementSibling = element.parentElement.nextElementSibling;

      const name = element.textContent;

      const slug = element.textContent
        .normalize('NFD')
        .replace(/[\u0300-\u036f]/g, '')
        .toLowerCase()
        .replace(/ /g, '-')
        .replace(/[^\w-]+/g, '');

      const products = Array.from(
        elementSibling.querySelectorAll(CONSTANTS.productCardClass),
      );

      return {
        name,
        slug,
        products: products.map((product, index) => {
          //TODO find a better way to deal with prices
          const value = product
            .querySelector(CONSTANTS.productMainPriceClass)
            .innerHTML.replace('R$ ', '')
            .replace('.', '');
          const decimal =
            product.querySelector(CONSTANTS.productDecimalPriceClass)
              ?.innerHTML || 0;

          const price = `${value}.${decimal}`;

          return {
            position: index + 1,
            name: product.querySelector(CONSTANTS.productNameClass).textContent,
            image: product
              .querySelector(CONSTANTS.productImageClass)
              .getAttribute('src'),
            price: +price,
            link: product.parentElement.getAttribute('href'),
          };
        }),
      };
    });
  }, CONSTANTS);

  await browser.close();

  return { categories: data };
};

File: src/main/java/com/laporeon/meliscraper/helpers/Scraper.java
package com.laporeon.meliscraper.helpers;

import com.laporeon.meliscraper.dtos.CategoryDTO;
import com.laporeon.meliscraper.dtos.ProductDTO;
import lombok.SneakyThrows;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.text.Normalizer;
import java.util.List;

@Component
public class Scraper {

    public static final String APP_URL = "https://www.mercadolivre.com.br/mais-vendidos";
    public static final String CATEGORIES_SECTION_CLASS = ".dynamic__carousel";
    public static final String CATEGORY_TITLE_CLASS = ".dynamic-carousel__container--with-link > h2";
    public static final String PRODUCT_CARD_CLASS = ".dynamic-carousel__item-container";
    public static final String PRODUCT_NAME_CLASS = ".dynamic-carousel__title";
    public static final String PRODUCT_IMAGE_CLASS = ".dynamic-carousel__img";
    public static final String PRODUCT_MAIN_PRICE_CLASS = ".dynamic-carousel__price > span";
    public static final String PRODUCT_DECIMAL_PRICE_CLASS = ".dynamic-carousel__price-decimals";
    public static final String PRODUCT_LINK_CLASS = ".dynamic-carousel__item-container > a";

    @SneakyThrows
    public List<CategoryDTO> scrape() {
        Document doc = Jsoup.connect(APP_URL).get();
        return doc.select(CATEGORIES_SECTION_CLASS)
                  .stream()
                  .map(this::extractCategory)
                  .toList();
    }

    private CategoryDTO extractCategory(Element section) {
        String title = section.select(CATEGORY_TITLE_CLASS).text();
        String slug = slugify(title);
        List<ProductDTO> products = extractProducts(section);

        return new CategoryDTO(title, slug, products);
    }

    private List<ProductDTO> extractProducts(Element section) {
        Elements productElements = section.select(PRODUCT_CARD_CLASS);

        return productElements.stream()
                              .map(product -> new ProductDTO(
                                      productElements.indexOf(product) + 1,
                                      product.select(PRODUCT_NAME_CLASS).text(),
                                      extractImage(product),
                                      extractPrice(product),
                                      product.select(PRODUCT_LINK_CLASS).attr("href"),
                                      null))
                              .toList();
    }

    private static String extractImage(Element product) {
        Element img = product.select(PRODUCT_IMAGE_CLASS).first();

        if (!img.attr("data-src").isEmpty()) {
            return img.attr("data-src");
        }

        return img.attr("src");

    }

    private static BigDecimal extractPrice(Element product) {
        String main = product.select(PRODUCT_MAIN_PRICE_CLASS)
                             .text()
                             .replaceAll("\\D", "");


        String decimal = product.select(PRODUCT_DECIMAL_PRICE_CLASS)
                                .stream()
                                .findFirst()
                                .map(e -> e.text().replaceAll("\\D", ""))
                                .orElse("00");

        return new BigDecimal(main + decimal).movePointLeft(2);
    }


    private static String slugify(String name) {
        return Normalizer.normalize(name, Normalizer.Form.NFD)
                         .replaceAll("\\p{InCombiningDiacriticalMarks}+", "")
                         .toLowerCase()
                         .replace(" ", "-")
                         .replaceAll("[^\\w-]+", "");
    }

}